dbus.go

```diff
@@ -0,0 +1,157 @@
+package main
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/godbus/dbus/v5"
+)
+
+// SecretService implements the freedesktop.org Secret Service API as needed
+// for git-credential-fdoss to do what it needs to do. It is not a full
+// implementation of the API.
+//
+// https://specifications.freedesktop.org/secret-service/latest
+type SecretService struct {
+	conn    *dbus.Conn
+	svc     dbus.BusObject
+	session dbus.BusObject
+}
+
+type Secret struct {
+	Session     dbus.ObjectPath
+	Params      []byte
+	Secret      []byte
+	ContentType string
+}
+
+func NewSecretService() (*SecretService, error) {
+	conn, err := dbus.SessionBus()
+	if err != nil {
+		return nil, fmt.Errorf("couldn't connect to session bus: %w", err)
+	}
+	svc := conn.Object("org.freedesktop.secrets", dbus.ObjectPath("/org/freedesktop/secrets"))
+
+	var path dbus.ObjectPath
+	var output dbus.Variant
+	call := svc.Call("org.freedesktop.Secret.Service.OpenSession", 0, "plain", dbus.MakeVariant(""))
+	if err := call.Store(&output, &path); err != nil {
+		return nil, fmt.Errorf("couldn't open secret session: %w", err)
+	}
+
+	session := conn.Object("org.freedesktop.secrets", path)
+
+	return &SecretService{
+		conn:    conn,
+		svc:     svc,
+		session: session,
+	}, nil
+}
+
+func (ss *SecretService) Close() error {
+	call := ss.session.Call("org.freedesktop.Secret.Session.Close", 0)
+	return call.Err
+}
+
+func (ss *SecretService) Get(attrs map[string]string) (string, error) {
+	unlocked, locked, err := ss.search(attrs)
+	if err != nil {
+		return "", err
+	}
+
+	switch {
+	case len(unlocked) > 1:
+		fmt.Fprintln(os.Stderr, "Warning: Got more than one secret back. Using only the first")
+		fallthrough
+	case len(unlocked) > 0:
+		secret, err := ss.getSecret(unlocked[0])
+		if err != nil {
+			return "", err
+		}
+		return string(secret.Secret), nil
+	case len(locked) > 0:
+		fmt.Fprintln(os.Stderr, "TODO: Found locked secret. Sorry, can't unlock yet")
+	default:
+		// secret not found. return nothing
```

Comment:
shouldn't be an error? (same above?)

Reply:
Nope - it's fine to not have any secrets stored - you just can't return any and git will then prompt for the credentials. That's essentially how they get in there in the first place. You can also potentially have multiple credential helpers that git will call, which it will do until it has the username and password, so a credential helper should not produce an error if it does not have credentials.